File: MTG Deck Editor/src/gui/ScrollablePanel.java
package gui;

import java.awt.Dimension;
import java.awt.LayoutManager;
import java.awt.Rectangle;

import javax.swing.JPanel;
import javax.swing.Scrollable;

/**
 * This class is a JPanel except it is scroll-savvy.
 * 
 * @author Alec Roelke
 * @see javax.swing.JPanel
 * @see javax.swing.JScrollPane
 * @see javax.swing.Scrollable
 */
@SuppressWarnings("serial")
public class ScrollablePanel extends JPanel implements Scrollable
{
	/**
	 * Track neither width nor height of the viewport.
	 */
	public static final byte TRACK_NEITHER = 0x00;
	/**
	 * Track width of the viewport.
	 */
	public static final byte TRACK_WIDTH = 0x01;
	/**
	 * Track height of the viewport.
	 */
	public static final byte TRACK_HEIGHT = 0x02;
	/**
	 * Track both width and height of the viewport (equivalent to TRACK_WIDTH | TRACK_HEIGHT).
	 */
	public static final byte TRACK_BOTH = 0x03;
	
	/**
	 * Dimension of this ScrollablePanel's viewport to track.
	 */
	private byte tracking;
	
	/**
	 * Create a new ScrollablePanel with the default layout for a JPanel that
	 * tracks neither the width nor height of its viewport.
	 */
	public ScrollablePanel()
	{
		this(TRACK_NEITHER);
	}
	
	/**
	 * Create a new ScrollablePanel with the default layout for a JPanel
	 * that tracks the specified dimension of its viewport.
	 * 
	 * @param t
	 */
	public ScrollablePanel(byte t)
	{
		super();
		tracking = t;
	}
	
	/**
	 * Create a new ScrollablePanel with the specified layout that tracks
	 * neither the width nor height of its viewport.
	 * 
	 * @param layout Layout of the new panel
	 */
	public ScrollablePanel(LayoutManager layout)
	{
		this(layout, TRACK_NEITHER);
	}
	
	/**
	 * Create a new ScrollablePanel with the specified layout that tracks
	 * the specified dimension of its viewport.
	 * 
	 * @param layout
	 * @param t
	 */
	public ScrollablePanel(LayoutManager layout, byte t)
	{
		super(layout);
		tracking = t;
	}
	
	/**
	 * @return The preferred viewport size of this ScrollablePanel, which is the same
	 * as its preferred size.
	 */
	@Override
	public Dimension getPreferredScrollableViewportSize()
	{
		return getPreferredSize();
	}

	/**
	 * @see Scrollable#getScrollableBlockIncrement(Rectangle, int, int)
	 */
	@Override
	public int getScrollableBlockIncrement(Rectangle visibleRect, int orientation, int direction)
	{
		return 20;
	}

	/**
	 * @return <code>false</code> since the size of this ScrollablePanel should not track the
	 * viewport height.
	 */
	@Override
	public boolean getScrollableTracksViewportHeight()
	{
		return (tracking&TRACK_HEIGHT) != 0x0;
	}

	/**
	 * @return <code>false</code> since the size of this ScrollablePanel should not track the
	 * viewport width.
	 */
	@Override
	public boolean getScrollableTracksViewportWidth()
	{
		return (tracking&TRACK_WIDTH) != 0x0;
	}

	/**
	 * @see Scrollable#getScrollableUnitIncrement(Rectangle, int, int)
	 */
	@Override
	public int getScrollableUnitIncrement(Rectangle visibleRect, int orientation, int direction)
	{
		return 20;
	}
}

File: src/main/scala/editor/gui/generic/ScrollablePanel.scala
package editor.gui.generic

import javax.swing.JPanel
import javax.swing.Scrollable
import java.awt.LayoutManager
import java.awt.Dimension
import java.awt.Rectangle

object ScrollablePanel {
  val TrackWidth: Byte = 0x01
  val TrackHeight: Byte = 0x02
  val TrackBoth: Byte = (TrackWidth | TrackHeight).toByte
  val TrackNeither: Byte = 0x00
}

class ScrollablePanel(tracking: Byte = ScrollablePanel.TrackNeither, layout: LayoutManager = JPanel().getLayout) extends JPanel(layout) with Scrollable {
  @deprecated def this(tracking: Byte) = this(tracking, JPanel().getLayout)

  private var preferredScrollableViewportSize = Dimension(0, 0)

  def setPreferredScrollableViewportSize(size: Dimension) = preferredScrollableViewportSize = size

  override def getPreferredScrollableViewportSize = preferredScrollableViewportSize
  override def getScrollableUnitIncrement(visibleRect: Rectangle, orientation: Int, direction: Int) = 20
  override def getScrollableBlockIncrement(visibleRect: Rectangle, orientation: Int, direction: Int) = 20
  override def getScrollableTracksViewportWidth = (tracking & ScrollablePanel.TrackWidth) != 0
  override def getScrollableTracksViewportHeight = (tracking & ScrollablePanel.TrackHeight) != 0
}